src/main/scala/it/pagopa/interop/backendforfrontend/api/impl/AgreementsApiServiceImpl.scala

```diff
@@ -519,6 +519,34 @@ final case class AgreementsApiServiceImpl(
     }
   }
 
+  def getAgreementConsumers(q: Option[String], offset: Int, limit: Int)(implicit
+    contexts: Seq[(String, String)],
+    toEntityMarshallerProblem: ToEntityMarshaller[Problem],
+    toEntityMarshallerCompactOrganizations: ToEntityMarshaller[CompactOrganizations]
+  ): Route = {
+    val result: Future[CompactOrganizations] =
+      for {
+        _            <- validateQueryName(q)
+        pagedResults <- agreementProcessService.getAgreementConsumers(q, offset = offset, limit = limit)
+      } yield CompactOrganizations(
+        results = pagedResults.results.map(t => CompactOrganization(id = t.id, name = t.name)),
+        pagination = Pagination(offset = offset, limit = limit, totalCount = pagedResults.totalCount)
+      )
+
+    onComplete(result) {
+      handleError(
+        s"Error retrieving consumers from agreement filtered by consumer name $q, offset $offset, limit $limit"
+      ) orElse {
+        case Failure(_: InvalidQueryParameter) =>
+          getAgreementConsumers200(
+            CompactOrganizations(results = Nil, pagination = Pagination(offset = offset, limit = limit, totalCount = 0))
+          )
```

Comment:
I don't like it, this is exception based programming. You can remove this function (if not used elsewhere)

```scala
  private def validateQueryName(q: Option[String]): Future[Unit] = {
    q match {
      case Some(value) if value.length < 3 => Future.failed(InvalidQueryParameter(value))
      case _                               => Future.successful(())
    }
  }
```

and rewrite the main for comprehension to something like

```scala
def emptyResponse: Future[CompactOrganizations] = Future.successful(CompactOrganizations(results = Nil, pagination = Pagination(offset = offset, limit = limit, totalCount = 0)))

val result: Future[CompactOrganizations] = q.filter(_.length > 3).fold(emptyResponse)(validQuery =>
  agreementProcessService.getAgreementConsumers(validQuery, offset = offset, limit = limit).map(pagedResults =>
    CompactOrganizations(
        results = pagedResults.results.map(t => CompactOrganization(id = t.id, name = t.name)),
        pagination = Pagination(offset = offset, limit = limit, totalCount = pagedResults.totalCount)
    )
  )
)
```

Also, unless needed elsewhere, you can get rid of `InvalidQueryParameter`.

We can fix it together if needed, just ping me and I'll be happy to explain it further.

Reply:
Amazing!